### PR TITLE
Pd 2118 update build indices pipeline

### PR DESCRIPTION
**Pull request by khajoue2**
* Updated modify_gtf.py script to include two methods for Refseq vs Gencode reference files
* Updated the docker and the pipeline 
* Trimmed the tasks to only build star tar reference for Optimus based on trimmed gtf file
* Added this pipeline to dockstore yaml file

### CODE REVIEW

File: pipelines/skylab/build_indices/BuildIndices.wdl [hunk 1]
```diff
@@ -106,365 +26,70 @@ task BuildStarSingleNucleus {
...
 workflow BuildIndices {
   input {
-    String gtf_version
-    String organism
-    String organism_prefix
-    String genome_short_string
-    String dbsnp_version
+    File annotations_gtf
+    File genome_fa
+    String gtf_version = "M31"
```
Comment (jessicaway) [inline pipelines/skylab/build_indices/BuildIndices.wdl:67]: Should we have a default value here? If yes, perhaps all the others should also be defaulted?

File: pipelines/skylab/build_indices/BuildIndices.wdl [hunk 2]
```diff
@@ -106,365 +26,70 @@ task BuildStarSingleNucleus {
     set -eo pipefail
 
     python3 /script/modify_gtf.py  \
-    --input-gtf ~{references.annotation_gtf} \
+    --input-gtf ~{annotation_gtf} \
     --output-gtf ~{annotation_gtf_modified} \
```
Comment (jessicaway) [inline pipelines/skylab/build_indices/BuildIndices.wdl:30]: Are these always named `"modified_gencode.v~{gtf_version}.primary_assembly.annotation.gtf"`?

Probably we should at minimum remove the "gencode" part

**Review by jessicaway — COMMENTED** [review]
Looks good overall! A couple pretty minor comments

**Review by jessicaway — APPROVED** [review]

File: pipelines/skylab/build_indices/Macaque.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+"BuildIndices.annotations_gtf":"gs://fc-c40ec8a8-d60f-42f7-be36-3986b475190a/Macaque/genomic.gtf",
+"BuildIndices.biotypes":"gs://fc-df68cb43-8c48-401b-9ef1-7cbb3acc788d/Biotypes.tsv",
+"BuildIndices.genome_fa":"gs://fc-c40ec8a8-d60f-42f7-be36-3986b475190a/Macaque/GCF_003339765.1_Mmul_10_genomic.fna",
```
Comment (nikellepetrillo) [inline pipelines/skylab/build_indices/Macaque.json:4]: i've never heard of a .fna file. is this just a regular fasta file? or was this a typo?

File: pipelines/skylab/build_indices/BuildIndices.wdl [hunk 4]
```diff
@@ -106,365 +26,70 @@ task BuildStarSingleNucleus {
...
-             genome_fa: references.genome_fa,
-             annotation_gtf: annotation_gtf_modified
-           }
+    File modified_annotation_gtf = annotation_gtf_modified
   }
-  
+
   runtime {
-    docker: "us.gcr.io/broad-gotc-prod/build-indices:1.0.0-2.7.10a-1663605340"
+    docker: "us.gcr.io/broad-gotc-prod/build-indices:1.0.0-2.7.10a-1671132408"
```
Comment (nikellepetrillo) [inline pipelines/skylab/build_indices/BuildIndices.wdl:54]: If this is the finalized version of the docker we are going to merge into develop, do you mind adding a line to the .tsv? In dockers/skylab/build-indices/docker_versions.tsv

**Review by nikellepetrillo — APPROVED** [review]
just a few small comments, but looks good!

File: dockers/skylab/build-indices/modify_gtf.py [hunk 5]
```diff
@@ -2,29 +2,42 @@
...
     for f in features.split(';'):
-        if f:
+        if f and (len(f.split())>1):
             key = f.split()[0]
             value = f.split()[1]
             if key not in features_dic:
                 features_dic[key] = value
             else:
-                if type(features_dic[key]) == list:
+                if type(features_dic[key])==list:
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply dockers/skylab/build-indices/modify_gtf.py:42]: You are right! Added them back.

File: dockers/skylab/build-indices/modify_gtf.py [hunk 6]
```diff
@@ -62,11 +90,11 @@ def get_gene_ids(input_gtf,biotypes):
...
             if (features_dic['gene_type'] not in biotypes) or (
-                features_dic['transcript_type'] not in biotypes):
+                    features_dic['transcript_type'] not in biotypes):
                 continue
             if 'tag' in features_dic:
                 if ('readthrough_transcript' not in features_dic['tag']) and (
-                    'PAR' not in features_dic['tag']):
+                        'PAR' not in features_dic['tag']):
                     gene=features_dic['gene_id'].split('.', 1)[0]
                     if gene not in gene_ids:
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply dockers/skylab/build-indices/modify_gtf.py:103]: Removed the space

File: dockers/skylab/build-indices/modify_gtf.py [hunk 7]
```diff
@@ -34,22 +47,37 @@ def get_features(features):
...
+    with open(input_gtf, 'r') as input_file:
+        for line in input_file:
+            if not line.startswith('#'):
+                fields = [x.strip() for x in line.strip().split('\t')]
+                if fields[2]=='gene':
+                    features = re.sub('"', '', line.strip().split('\t')[8].strip())
+                    features_dic = get_features(features)
+                    if ('gene_biotype' in features_dic.keys()) and (features_dic['gene_biotype'] in biotypes):
+                        gene=features_dic['gene_id'].split('.',1)[0]
+                        if gene not in gene_ids:
```
[the remark this answers is not in the document]
Comment (khajoue2) [reply dockers/skylab/build-indices/modify_gtf.py:79]: Thank you! This is was the right thing to change and it made it much faster!

File: pipelines/skylab/build_indices/Macaque.json (hunk 3 above)
Comment (khajoue2) [reply pipelines/skylab/build_indices/Macaque.json:4]: The extension *.fna just means fasta nucleic acid while *.fa or *.fasta formats contain other DNA-related information

File: pipelines/skylab/build_indices/BuildIndices.wdl (hunk 1 above)
Comment (khajoue2) [reply pipelines/skylab/build_indices/BuildIndices.wdl:67]: removed the default

File: pipelines/skylab/build_indices/BuildIndices.wdl (hunk 2 above)
Comment (khajoue2) [reply pipelines/skylab/build_indices/BuildIndices.wdl:30]: Removed gencode and primary.assembly

File: pipelines/skylab/build_indices/BuildIndices.wdl (hunk 4 above)
Comment (khajoue2) [reply pipelines/skylab/build_indices/BuildIndices.wdl:54]: No, I just updated the docker with changes Jess recommended and added the a new tag to the wdl. Thank you for reminding me to update the tsv!

**Review by ekiernan — APPROVED** [review]

**Pull request merged**